compiler/luci/pass/src/FuseSliceWithTConvPass.cpp

```diff
@@ -0,0 +1,286 @@
+ *                    |
+ *            [CircleTransposeConv] (with m.b. changed padding, output shape, and filter/bias)
+ *                    |
+ *
+ */
+
+bool fuse_slice_with_tconv(luci::CircleSlice *slice)
+{
+  // NOTE: assume NHWC layout
+  auto tconv = dynamic_cast<luci::CircleTransposeConv *>(slice->input());
+  RETURN_FALSE_UNLESS(tconv != nullptr);
+
+  auto const tconv_users = loco::succs(tconv);
+  // in general no fusion is possible if tconv is used by many successors
+  RETURN_FALSE_UNLESS(tconv_users.size() == 1);
+
+  // offset
+  auto begin = dynamic_cast<luci::CircleConst *>(slice->begin());
+  // sanity check
+  RETURN_FALSE_UNLESS(begin != nullptr && begin->dtype() == loco::DataType::S32 &&
+                      begin->rank() == 1);
+
+  // output shape
+  auto size = dynamic_cast<luci::CircleConst *>(slice->size());
+  // sanity check
+  RETURN_FALSE_UNLESS(size != nullptr && size->dtype() == loco::DataType::S32 && size->rank() == 1);
+
+  // output shape of tconv
+  auto tconv_shape = dynamic_cast<luci::CircleConst *>(tconv->inputSizes());
+  // sanity check
+  RETURN_FALSE_UNLESS(tconv_shape != nullptr && tconv_shape->dtype() == loco::DataType::S32 &&
+                      tconv_shape->rank() == 1);
+
+  // no update if batch dimension is processed in slice
+  RETURN_FALSE_UNLESS(begin->at<loco::DataType::S32>(0) == 0 &&
+                      size->at<loco::DataType::S32>(0) == tconv_shape->at<loco::DataType::S32>(0));
+
+  auto const out_height = size->at<loco::DataType::S32>(1);
+  auto const out_width = size->at<loco::DataType::S32>(2);
+
+  auto const pad_top = begin->at<loco::DataType::S32>(1);
+  auto const pad_left = begin->at<loco::DataType::S32>(2);
+
+  // As there is no option to set numerical values of pad explicitly for CircleTransposeConv
+  // we need to be sure that interpretation of PADDING + OUTPUT_SHAPE will produce
+  // the pad values, defined by slice. If possible compute_padding will return correct
+  // padding value, otherwise it will return UNDEFINED
+  auto const padding = compute_padding(tconv, out_height, out_width, pad_top, pad_left);
+  if (padding == luci::Padding::UNDEFINED)
+    return false; // no fuse is possible
+
+  auto const out_channels = size->at<loco::DataType::S32>(3);
+  // update filter and bias in case it's needed
+  loco::Node *fused_filter = tconv->filter();
+  loco::Node *fused_bias = tconv->bias();
+  if (begin->at<loco::DataType::S32>(3) != 0 ||
+      out_channels != tconv_shape->at<loco::DataType::S32>(3))
+  {
+    auto const tconv_filter = dynamic_cast<luci::CircleConst *>(tconv->filter());
+    // sanity check
+    RETURN_FALSE_UNLESS(tconv_filter != nullptr && tconv_filter->rank() == 4 &&
+                        tconv_filter->dtype() == loco::DataType::FLOAT32);
+
+    auto const in_channels = tconv_filter->dim(3).value();
+
+    luci::CircleConst *fused_tconv_filter =
+      loco::must_cast<luci::CircleConst *>(luci::clone_node(tconv_filter, slice->graph()));
```

Comment:
Node is created in this line, meaning that graph is changed. But in the below line, this function can return false, e.g., if bias is non-const.

Could you check all conditions first and then create nodes?

Reply:
ok